teps/0142-enable-step-reusability.md

```diff
@@ -0,0 +1,997 @@
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: ci-t
+spec:
+  workspaces:
+    - name: shared-data
+  params:
+    - name: repo-url
+    - name: revision
+    - name: image-url
+    - name: dockerfile
+  results:
+    - name: commit
+    - name: digest
+  steps:
+    - uses:
+        path: tektoncd/catalog/task/git-clone/0.9/git-clone.yaml@HEAD
+    - uses:
+        path: tektoncd/catalog/task/git-clone/0.4/kaniko.yaml@HEAD
+```
+
+#### Comparable CI/CD Systems
+
+##### 1. Github Actions
+
+[GitHub Actions](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsuses) allows users to specify `Actions` to run as part of a `Step` in a `Job`. Note that `Action` is the reusable unit of work, and it can contain multiple `Steps`. The `Actions` can execute in an environment with a shared file system, so the users don’t have to choose between reusability and performance.
+
+```yaml
+jobs:
+  clone-build-push:
+    steps:
+    - uses: actions/checkout@v0.9
+      with:
+        repo-url: github.com/foo/bar.git
+        revision: main
+    - uses: actions/kaniko@v0.4
+      with:
+        image-url: gcr.io/my_app:version
+        dockerfile: ./Dockerfile
+```
+
+## Proposal
+
+We propose introducing a `StepAction` CRD (i.e. “the work performed by the Step” ) composed of the actionable parts of the current `Step` struct. See the [Step CRD](#stepcrd) section for why we did not pick this name instead.
+
+The `StepAction` will be the reusable and scriptable unit of work that is defined at authoring time and does not contain runtime components.
+
+**Note**: A `Step` is not reusable, the work ir performs is reusable and referenceable. `Steps` are in-lined in the `Task` definition and either perform work directly or perform a `StepAction`. A `StepAction` cannot be run stand-alone (unlike a `TaskRun` or a `PipelineRun`). It has to be referenced by a `Step`. In other words, a `Step` is not composed of `StepActions` (unlike a `Task` being composed of `Steps` and `Sidecars`). A `Step` is an actionable component, meaning that it has the ability to refer to a `StepAction`. The `Task` author should be able to compose a `Step` using a `StepAction` and provide all the necessary context (or orchestration) to it.` 
+
+With this solution, reusable units of work can execute in the same environment with a shared filesystem which improves usability and security. It also eliminates the unnecessary creation of new environments and enables use of local disks to optimize performance.
+
+### StepAction CRD
+The `Step` struct can be split into “actionable” (aka. “work”) and “orchestration” fields. 
```

Comment:
This is a nit, but we use `struct` here, which makes little sense when talking about API (`struct` is a Go concept). We should use "object" or something.

*The reason for this is, technically the API could be written in Rust or anything*.